drivers/eeprom/eeprom_emulator.c

```diff
@@ -0,0 +1,543 @@
+/*
+ * Copyright (c) 2020 Laczen
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * This driver emulates a EEPROM device in flash.
+ *
+ * The emulation represents the EEPROM in flash as a region that is a direct
+ * map of the eeprom data followed by a region where changes to the eeprom
+ * data is stored. Changes are written as address-data combinations. The size
+ * of such a combination is determined by the flash write block size and
+ * the size of the eeprom (required address space), with a minimum of 4 byte.
+ * The eeprom page needs to be a multiple of the flash page. Multiple eeprom
+ * pages is also so supported and increases the number of writes that can be
+ * performed.
+ *
+ * Internally the address of an eeprom byte is represented by a uint32_t (this
+ * should be sufficient in all cases). In case the eeprom size is smaller than
+ * 64kB only a uint16_t is used to store changes. In this case the change stored
+ * for a 4 byte flash write block size are a combination of 2 byte address and
+ * 2 byte data.
+ *
+ * The eeprom size, pagesize and the flash partition used for the eeprom are
+ * defined in the dts. The flash partition should allow at least two eeprom
+ * pages.
+ *
+ */
+
+#define DT_DRV_COMPAT zephyr_emu_eeprom
+#include <drivers/eeprom.h>
+#include <drivers/flash.h>
+#include <zephyr.h>
+#define LOG_LEVEL CONFIG_EEPROM_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(eeprom_emulator);
+
+#define DEV_CONFIG(dev) ((dev)->config)
+#define DEV_DATA(dev) ((dev)->data)
+
+
+struct eeprom_emu_config {
+	size_t size;
+	bool readonly;
+	size_t page_size;
+	off_t flash_offset;
+	size_t flash_size;
+	uint8_t flash_cbs;
+	const struct device *flash_dev;
+};
+
+struct eeprom_emu_data {
+	off_t write_offset;
+	off_t page_offset;
+	struct k_mutex lock;
+};
+
+/*
+ * basic flash read, only used with offset aligned to flash write block size
+ */
+static void eeprom_emu_flash_read(const struct device *dev, off_t offset,
+				  uint8_t *blk, size_t len)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+
+	if (device_is_ready(dev_config->flash_dev)) {
+		offset += dev_config->flash_offset;
+		if (!flash_read(dev_config->flash_dev, offset, blk, len)) {
+			return;
+		}
+	}
+	__ASSERT(0, "Flash interface failed");
+}
+
+/*
+ * basic flash write, only used with offset aligned to flash write block size
+ */
+static void eeprom_emu_flash_write(const struct device *dev, off_t offset,
+				   const uint8_t *blk, size_t len)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+	int rc;
+
+	if (device_is_ready(dev_config->flash_dev)) {
+		offset += dev_config->flash_offset;
+		rc = flash_write_protection_set(dev_config->flash_dev, false);
+		rc |= flash_write(dev_config->flash_dev, offset, blk, len);
+		rc |= flash_write_protection_set(dev_config->flash_dev, true);
+		if (!rc) {
+			return;
+		}
+	}
+	__ASSERT(0, "Flash interface failed");
```

Comment:
Because it should also assert if there is a flash error. 
I am trying to rework the flash write/read/erase routines to allow better error return codes, this however is not so easy because the eeprom_api returns `int` instead of `ssize_t`. And an api change is not on my to do list. An idea ?